File: scraper.js
// Import required modules
const fs = require('fs');
const readline = require('readline');
const puppeteer = require('puppeteer-extra');
const StealthPlugin = require('puppeteer-extra-plugin-stealth');
puppeteer.use(StealthPlugin());

// Main scraping function
async function scrapeTikTok(username) {
  const url = `https://www.tiktok.com/@${username}`;
  const browser = await puppeteer.launch({ headless: false, args: ['--no-sandbox'] });
  const page = await browser.newPage();

  await page.goto(url, { waitUntil: 'networkidle2' });
  console.log(`🔍 Scraping profile: ${url}`);

  await page.waitForSelector('h1[data-e2e="user-title"]');

  const profile = await page.evaluate(() => {
    const getText = (selector) => {
      const el = document.querySelector(selector);
      return el ? el.textContent.trim() : null;
    };

    return {
      username: getText('h1[data-e2e="user-title"]'),
      displayName: getText('h2[data-e2e="user-subtitle"]'),
      bio: getText('h2[data-e2e="user-bio"]'),
      followers: getText('strong[data-e2e="followers-count"]'),
      following: getText('strong[data-e2e="following-count"]'),
      likes: getText('strong[data-e2e="likes-count"]')
    };
  });

  console.log('✅ Profile scraped:\n', profile);

  try {
    await page.waitForSelector('button.TUXSegmentedControl-item', { timeout: 7000 });
    const tabButtons = await page.$$('button.TUXSegmentedControl-item');
    let popularButton = null;
    for (const btn of tabButtons) {
      const text = await btn.evaluate(el => el.textContent.trim());
      if (text === 'Popular') {
        popularButton = btn;
        break;
      }
    }
    if (!popularButton) throw new Error('Popular tab button not found');
    const isActive = await popularButton.evaluate(el => el.getAttribute('data-active') === 'true');
    if (!isActive) {
      await popularButton.click();
      console.log('➡️ Switched to "Popular" tab, waiting for videos to load...');
      await delay(3000);
    } else {
      console.log('ℹ️ Already on "Popular" tab');
    }
  } catch (e) {
    console.warn(`⚠️ Failed to switch to "Popular" tab: ${e.message}`);
  }

  await autoScroll(page);

  const popularVideos = await page.$$eval('div[data-e2e="user-post-item"] a', links => {
    return links.map(a => {
      const container = a.closest('div[data-e2e="user-post-item"]');
      const viewEl = container?.querySelector('strong[data-e2e="video-views"]');
      return {
        url: a.href,
        views: viewEl ? viewEl.textContent.trim() : '0'
      };
    }).filter(v => v.url.includes('/video/'));
  });

  console.log(`📦 Total videos available in Popular tab: ${popularVideos.length}`);

  const limit = await askVideoLimit(popularVideos.length);
  const videos = [];
  const videoPage = await browser.newPage();

  for (let i = 0; i < limit; i++) {
    const videoUrl = popularVideos[i].url;
    const extractedViews = popularVideos[i].views;

    try {
      console.log(`🎥 Scraping video ${i + 1}/${limit}: ${videoUrl}`);
      await videoPage.goto(videoUrl, { waitUntil: 'networkidle2', timeout: 60000 });
      await videoPage.waitForSelector('[data-e2e="like-count"], span[data-e2e="like-count"]', { timeout: 20000 });

      const videoData = await videoPage.evaluate(() => {
        const getTextByAttr = (attrVal) => {
          const el = document.querySelector(`[data-e2e="${attrVal}"]`);
          return el ? el.textContent.trim() : '0';
        };

        const getTextByFallback = (query) => {
          const el = document.querySelector(query);
          return el ? el.textContent.trim() : '0';
        };

        return {
          likes: getTextByAttr('like-count') || getTextByFallback('span[data-e2e="like-count"]'),
          comments: getTextByAttr('comment-count') || getTextByFallback('span[data-e2e="comment-count"]'),
          shares: getTextByAttr('share-count') || getTextByFallback('span[data-e2e="share-count"]'),
        };
      });

      videos.push({
        videoLink: videoUrl,
        likes: videoData.likes,
        comments: videoData.comments,
        shares: videoData.shares,
        views: extractedViews
      });

      console.log(`✅ Scraped: ${videoData.likes} 👍 | ${videoData.comments} 💬 | ${videoData.shares} 🔁 | ${extractedViews} 👀`);
      await delay(2000);
    } catch (err) {
      console.log(`⚠️ Failed to scrape video stats for ${videoUrl}:`, err.message);
    }
  }

  await videoPage.close();
  await browser.close();

  const result = { profile, totalPopularVideos: popularVideos.length, videos };
  fs.writeFileSync(`tiktok_${username}_popular.json`, JSON.stringify(result, null, 2));
  console.log(`📁 Results saved to tiktok_${username}_popular.json`);
  return result;
}

// Prompt user to input how many videos to fetch
async function askVideoLimit(max) {
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  const question = (str) => new Promise(resolve => rl.question(str, resolve));

  let limit;
  while (true) {
    const answer = await question(`🔢 Enter number of videos to scrape (max ${max}): `);
    limit = parseInt(answer);
    if (!isNaN(limit) && limit > 0 && limit <= max) break;
    console.log(`❌ Invalid input. Please enter a number between 1 and ${max}`);
  }

  rl.close();
  return limit;
}

// Ask user for TikTok username
async function askUsername() {
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
  const question = (str) => new Promise(resolve => rl.question(str, resolve));
  const username = await question('👤 Enter TikTok username (without @): ');
  rl.close();
  return username.trim();
}

// Cleaned dynamic auto-scroll function
async function autoScroll(page, maxScrolls = 50, pauseTime = 1000) {
  let lastVideoCount = 0;
  let unchangedScrolls = 0;

  for (let i = 0; i < maxScrolls; i++) {
    const currentCount = await page.$$eval('div[data-e2e="user-post-item"]', els => els.length);

    if (currentCount === lastVideoCount) {
      unchangedScrolls++;
    } else {
      unchangedScrolls = 0;
    }

    if (unchangedScrolls >= 3) {
      console.log('🛑 No new videos loaded after multiple scrolls. Stopping scroll...');
      break;
    }

    lastVideoCount = currentCount;
    await page.evaluate(() => window.scrollBy(0, window.innerHeight));
    await delay(pauseTime);
  }

  console.log(`✅ Finished scrolling. Total videos loaded: ${lastVideoCount}`);
}

// Delay helper
function delay(time) {
  return new Promise(resolve => setTimeout(resolve, time));
}

// Script entry point
(async () => {
  try {
    const username = await askUsername();
    const result = await scrapeTikTok(username);
    console.log('\n🎉 Scraping complete!');
    console.log('Profile info:', result.profile);
    console.log(`Total Popular Videos: ${result.totalPopularVideos}`);
    console.log('Top videos scraped:', result.videos.length);
  } catch (error) {
    console.error('❌ Scrape failed:', error);
  }
})();

File: scraper.py
import json
import asyncio
from playwright.async_api import async_playwright


async def delay(seconds):
    await asyncio.sleep(seconds)


async def ask_username():
    return input("👤 Enter TikTok username (without @): ").strip()


def ask_video_limit(max_limit):
    while True:
        try:
            val = int(input(f"🔢 Enter number of videos to scrape (max {max_limit}): "))
            if 1 <= val <= max_limit:
                return val
        except ValueError:
            pass
        print(f"❌ Invalid input. Please enter a number between 1 and {max_limit}")


async def auto_scroll(page, max_scrolls=50, pause_time=1.0):
    last_video_count = 0
    unchanged_scrolls = 0

    for _ in range(max_scrolls):
        current_count = await page.locator('div[data-e2e="user-post-item"]').count()
        if current_count == last_video_count:
            unchanged_scrolls += 1
        else:
            unchanged_scrolls = 0
        if unchanged_scrolls >= 3:
            print("🛑 No new videos loaded. Stopping scroll.")
            break
        last_video_count = current_count
        await page.evaluate("window.scrollBy(0, window.innerHeight);")
        await delay(pause_time)
    print(f"✅ Finished scrolling. Total videos loaded: {last_video_count}")


async def scrape_tiktok(username):
    url = f"https://www.tiktok.com/@{username}"

    async with async_playwright() as p:
        browser = await p.chromium.launch(headless=False, args=["--no-sandbox"])
        page = await browser.new_page()
        await page.goto(url, wait_until='networkidle')
        print(f"🔍 Scraping profile: {url}")

        await page.wait_for_selector('h1[data-e2e="user-title"]', timeout=15000)

        profile = {
            "username": await page.locator('h1[data-e2e="user-title"]').inner_text(),
            "displayName": await page.locator('h2[data-e2e="user-subtitle"]').inner_text(),
            "bio": await page.locator('h2[data-e2e="user-bio"]').inner_text(),
            "followers": await page.locator('strong[data-e2e="followers-count"]').inner_text(),
            "following": await page.locator('strong[data-e2e="following-count"]').inner_text(),
            "likes": await page.locator('strong[data-e2e="likes-count"]').inner_text()
        }
        print("✅ Profile scraped:\n", profile)

        # Attempt to switch to "Popular" tab
        try:
            await page.wait_for_selector('button.TUXSegmentedControl-item', timeout=7000)
            buttons = page.locator('button.TUXSegmentedControl-item')
            for i in range(await buttons.count()):
                text = await buttons.nth(i).inner_text()
                if text.strip().lower() == "popular":
                    is_active = await buttons.nth(i).get_attribute("data-active")
                    if is_active != "true":
                        await buttons.nth(i).click()
                        print("➡️ Switched to 'Popular' tab.")
                        await delay(3)
                    else:
                        print("ℹ️ Already on 'Popular' tab.")
                    break
        except Exception as e:
            print(f"⚠️ Could not switch to 'Popular' tab: {e}")

        await auto_scroll(page)

        anchors = page.locator('div[data-e2e="user-post-item"] a')
        video_count = await anchors.count()
        popular_videos = []

        for i in range(video_count):
            href = await anchors.nth(i).get_attribute('href')
            view_el = await anchors.nth(i).locator('xpath=../../..//strong[data-e2e="video-views"]').element_handle()
            views = await view_el.inner_text() if view_el else "0"
            if "/video/" in href:
                popular_videos.append({ "url": f"https://www.tiktok.com{href}", "views": views })

        print(f"📦 Found {len(popular_videos)} videos in 'Popular' tab.")
        limit = ask_video_limit(len(popular_videos))

        videos = []
        video_page = await browser.new_page()

        for i in range(limit):
            video = popular_videos[i]
            print(f"🎥 Scraping video {i + 1}/{limit}: {video['url']}")
            try:
                await video_page.goto(video["url"], wait_until='networkidle')
                await video_page.wait_for_selector('[data-e2e="like-count"]', timeout=15000)

                likes = await video_page.locator('[data-e2e="like-count"]').inner_text()
                comments = await video_page.locator('[data-e2e="comment-count"]').inner_text()
                shares = await video_page.locator('[data-e2e="share-count"]').inner_text()

                videos.append({
                    "videoLink": video["url"],
                    "likes": likes,
                    "comments": comments,
                    "shares": shares,
                    "views": video["views"]
                })
                print(f"✅ Scraped: {likes} 👍 | {comments} 💬 | {shares} 🔁 | {video['views']} 👀")
                await delay(2)
            except Exception as e:
                print(f"⚠️ Failed to scrape {video['url']}: {e}")

        await video_page.close()
        await browser.close()

        result = {
            "profile": profile,
            "totalPopularVideos": len(popular_videos),
            "videos": videos
        }

        filename = f"tiktok_{username}_popular.json"
        with open(filename, "w", encoding="utf-8") as f:
            json.dump(result, f, indent=2, ensure_ascii=False)

        print(f"📁 Results saved to {filename}")
        return result


if __name__ == "__main__":
    try:
        username = asyncio.run(ask_username())
        result = asyncio.run(scrape_tiktok(username))
        print("\n🎉 Scraping complete!")
        print("Profile info:", result["profile"])
        print(f"Total Popular Videos: {result['totalPopularVideos']}")
        print(f"Top videos scraped: {len(result['videos'])}")
    except Exception as e:
        print("❌ Scrape failed:", e)
